Add download button to the single audio result page

obulat opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #270 by @zackkrida

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR adds `DownloadButton` component to the Single audio result page: 

![download_audio](https://user-images.githubusercontent.com/15233243/135868375-54f0dae2-531a-4536-804c-8d75ef799a77.gif)

Currently, the file extension is being set in the Catalog as `mp3`. However, Jamendo provides two types of `mp3` of different quality. Should we also set the filetype as mp31 and mp32 as Jamendo does?


## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

sarayourfriend commented:
> Should we also set the filetype as mp31 and mp32 as Jamendo does

This is where I think we still need to talk about how to store the file format name. I separated the format name from the URL because I assumed the true format name couldn't be derived from some download URLs (this is an example). We should have an MP3 96kbs and an MP3 V0 for jamendo for each of their available MP3 download formats.

sarayourfriend commented on src/components/AudioTrack/layouts/Full.vue:
```diff
@@ -52,9 +54,30 @@ export default {
+    const getFormats = (audio) => {
+      // TODO: finalize the property names in
+      //  formats objects and alt_files objects
+      let formats = [
+        { extension_name: audio.filetype, download_url: audio.url },
+      ]
```
If the `audio` object already uses `filetype` and `url` for the properties I think we should just update the download button component to accept those properties, so we can just pass a list of audio objects instead.

The only concern is whether `filetype` is a truly human readable filetype (like it needs to be `MP3 96kbs` instead of `mp31` for example)

zackkrida replied:
Yeah we might want to keep an enum of human readable file name mappings here, or more likely push this down to the catalog level and prettify the `format` before saving the rows.

obulat replied:
>  I separated the format name from the URL because I assumed the true format name couldn't be derived from some download URLs (this is an example). We should have an MP3 96kbs and an MP3 V0 for jamendo for each of their available MP3 download formats.

There is no _standard_ way of deriving the audio format from the URL: for example, Jamendo has format information after the last `/`, and Wikimedia often has it as an extension on the download URL. This is why we want to add `filetype` as a column to the Catalog database.

sarayourfriend commented on src/data/sample-audio-responses.json:
```diff
@@ -250,6 +252,14 @@
       "name": "Distance & Temps",
       "url": "https://www.jamendo.com/album/1290/distance-and-temps"
     },
+    "alt_files": [{
+      "extension": "flac",
+      "url": "https://mp3d.jamendo.com/download/track/8561/flac"
+      },
+      {
+      "extension": "mp31",
+      "url": "https://mp3d.jamendo.com/download/track/8561/mp31"
+    }],
```
Why use `extension` here instead of `filetype` to be consistent with the main result?

zackkrida replied:
yep, might be best to match the catalog fields, no?

https://github.com/WordPress/openverse-catalog/blob/a63fdf81755632f9c6ba83c822fd0fe8eef853a6/openverse_catalog/dags/provider_api_scripts/freesound.py#L198-L204

obulat replied:
This is the main reason why this PR is a draft: I wanted to discuss and finalized the naming and values for `extension` ( filetype?, human-readable-filetype?). I used `extension` for now because that is what's used in the `DownloadButton` component.
So, to make the naming consistent, I guess we could use `filetype` and `url` properties for main audio, and change the [`alt_files`](https://github.com/WordPress/openverse-catalog/blob/a63fdf81755632f9c6ba83c822fd0fe8eef853a6/openverse_catalog/dags/provider_api_scripts/freesound.py#L198-L204) to have the same properties (filetype instead of format):
```
alt_files: {
        'url': media_data.get('download'),
        'filetype': media_data.get('type'),
        'filesize': media_data.get('filesize'),
        'bit_rate': media_data.get('bitrate'),
        'sample_rate': media_data.get('samplerate'),
}

Then I think it would be easier to have a technical name in the catalog (ie. `mp31`, `mp32`), and convert it to a human-readable form on the front end. So, the front end would have a mapping to convert `mp31` coming from the API to `MP3 96kbs` for display on the button.

obulat replied:
I've opened a discussion about this: https://github.com/WordPress/openverse/discussions/85

sarayourfriend reviewed (commented):
Sorry for the early review but just some preliminary comments and thoughts. Otherwise I'm hopeful that this looks like it shouldn't be too bad to implement using the component 😁

zackkrida commented on src/components/AudioTrack/layouts/Full.vue:
```diff
@@ -28,8 +28,10 @@
-      <!-- TODO: Download dropdown -->
-      <div class="right-content bg-pink h-12 w-30 rounded-sm ml-auto" />
+      <DownloadButton
+        class="right-content h-14 flex"
```
```suggestion 
```

These don't seem to have any affect on the styling. Removing them doesn't change the appearance

obulat replied:
I agree. I thought that `left-content` and `right-content` were left in their for some styling later on, so I decided to leave it in.

zackkrida commented on src/components/DropdownButton.vue:
```diff
@@ -11,7 +11,7 @@
         ref="dropdownButton"
         type="button"
-        class="dropdown-button ml-1 rounded-r-sm rounded-l-none"
+        class="dropdown-button ml-1 rounded-r-sm rounded-l-none w-14"
```
I don't love the idea of a hardcoded width inside of this component, I'd rather see this be `w-full` and do the `w-14` on a parent at the page, implementation-level.

sarayourfriend replied:
I think a fixed width for the dropdown arrow button is fine... the download button should expand to fill the remaining space (`w-full` I guess?). I tried to implement that with flex but I guess I didn't do it right? I don't think we need to actually set explicit widths on anything? Except maybe a min-width on the download button in case the word "download" is very short in any given language.

zackkrida replied:
Oh i totally lazily missed that this was on the dropdown arrow button, oops. I agree with your thoughts @sarayourfriend

zackkrida commented:
Hm, it looks like a visual regression where the dropdown "drawer" _covers_ the button, instead of appearing below it.

obulat commented:
> Hm, it looks like a visual regression where the dropdown "drawer" _covers_ the button, instead of appearing below it.
> 
> CC @sarayourfriend would appreciate 👀 here

The regression was caused by changing the `DropdownButton`'s outer container's `display` property from `block` to `flex`, so that it takes up the 100% of parent's height, and becomes the same height as the play-pause button on the left. 
To fix it, I set the `top` property for the dropdown to the height of the play-pause and arrow buttons. But I used a CSS variable set on the `Full` audio track component, because the actual `DropdownButton` that controls the dropdown height is a generic-ish component, and it shouldn't have a set top offset for it. What is the best way of doing it using Tailwind?

Here is the result:
![download_audio](https://user-images.githubusercontent.com/15233243/136642509-62152f9f-3b7c-405d-b222-777844bb99c7.gif)

sarayourfriend commented:
> The regression was caused by changing the DropdownButton's outer container's display property from block to flex, so that it takes up the 100% of parent's height, and becomes the same height as the play-pause button on the left.

This is totally naive question, but is it possible to keep it as `block` but use `h-full` to make it fill the parent container's height? Not sure if that's possible or if it works.

obulat commented:
Copying the message from a PR to a PR :)
I've added some changes:
- made the Full layout top-aligned to prevent button moving downwards on a narrow screen, and adjusted the h1's line height so that it looks more center-aligned anyways. 
- added flex-shrink-0 to the download button so that it stays the same width.
- adjusted padding and line-height (leading-) on the download button so that it's height more or less matches the height of the download button. 

Does it look close to what you envisioned, @panchovm, @sarayourfriend ?
<img width="989" alt="Screen Shot 2021-10-15 at 2 28 05 PM" src="https://user-images.githubusercontent.com/15233243/137480340-4f4fbd60-09f4-48e5-b9bd-651b4a123478.png">
<img width="1013" alt="Screen Shot 2021-10-15 at 2 27 46 PM" src="https://user-images.githubusercontent.com/15233243/137480342-1e55d12b-cb9a-4192-ba12-c0944040851e.png">
